naga/src/back/msl/writer.rs

```diff
@@ -3720,6 +3730,734 @@ impl<W: Write> Writer<W> {
+                                                        (float(b2 << 8 | b3) - 32767.0f) / 65535.0f, \
+                                                        (float(b4 << 8 | b5) - 32767.0f) / 65535.0f, \
+                                                        (float(b6 << 8 | b7) - 32767.0f) / 65535.0f);",
+                                back::INDENT
+                            )?;
+                            writeln!(self.out, "}}")?;
+                            (name, 8)
+                        }
+                        Float16x2 => {
+                            let name = self.namer.call("unpackFloat16x2");
+                            writeln!(
+                                self.out,
+                                "metal::half2 {name}(metal::ushort b0, \
+                                                     metal::ushort b1, \
+                                                     metal::ushort b2, \
+                                                     metal::ushort b3) {{"
+                            )?;
+                            writeln!(
+                                self.out,
+                                "{}return metal::half2(as_type<metal::half>(b0 << 8 | b1), \
+                                                       as_type<metal::half>(b2 << 8 | b3));",
+                                back::INDENT
+                            )?;
+                            writeln!(self.out, "}}")?;
+                            (name, 4)
+                        }
+                        Float16x4 => {
+                            let name = self.namer.call("unpackFloat16x4");
+                            writeln!(
+                                self.out,
+                                "metal::half4 {name}(metal::ushort b0, \
+                                                     metal::ushort b1, \
+                                                     metal::ushort b2, \
+                                                     metal::ushort b3, \
+                                                     metal::ushort b4, \
+                                                     metal::ushort b5, \
+                                                     metal::ushort b6, \
+                                                     metal::ushort b7) {{"
+                            )?;
+                            writeln!(
+                                self.out,
+                                "{}return metal::half4(as_type<metal::half>(b0 << 8 | b1), \
+                                                       as_type<metal::half>(b2 << 8 | b3), \
+                                                       as_type<metal::half>(b4 << 8 | b5), \
+                                                       as_type<metal::half>(b6 << 8 | b7));",
+                                back::INDENT
+                            )?;
+                            writeln!(self.out, "}}")?;
+                            (name, 8)
+                        }
```

Comment:
I think the return type of all unpacking functions need to be 32-bit numbers (or vectors of 32-bit numbers).

From: https://gpuweb.github.io/gpuweb/#vertex-processing & https://gpuweb.github.io/gpuweb/wgsl/#channel-formats

WGSL without the `f16` extension (which we haven't implemented) only supports 32-bit numbers.

Reply:
But at this point, all the types and math are being done in MSL, using MSL types. wgsl types shouldn't matter anymore.

Reply:
Re-open if I'm missing something, but hopefully any problems here will be revealed by tests.

Reply:
I see the problem now, sorry, yes. Since Naga is generating all the MSL, and it only generates int, uint and float types (and vector variants), there's no value in these unpacking functions returning anything else. It also causes errors with Metal validation. For example, you can't assign a char4 to an int4. I'll fix this.